[#92231] Add time to ongoing reservation

jdzak opened the pull request:
https://pm.tablexi.com/issues/92231

jhanggi commented on app/controllers/reservations_controller.rb:
```diff
@@ -183,6 +183,15 @@ def update
       return
     end
 
+    if !current_user.administrator?
```
It seems like this should come after the `assign_times_from_params` call. I could be wrong, but without that, it probably won't maintain your changes you made through the interface.

jhanggi replied:
In the context where this action is used, is this check necessary? Most of the time, won't the administrator will be using one of the other reservation change views (like the order detail edit popup).

jdzak replied:
Yea, you're right about `assign_times_from_params`, I'll fix that.

Also, I'll remove the admin check for the validation. Sorry, I'm still kind of unfamiliar with how each role edits different details, but that makes sense.

jhanggi commented on app/controllers/reservations_controller.rb:
```diff
@@ -183,6 +183,15 @@ def update
+    if !current_user.administrator?
+      validator = Reservations::DurationChangeValidations.new(@reservation)
+      if validator.invalid?
+        validator.copy_errors!
```
Can we move this into the validator class itself?

jdzak replied:
Sure.

jhanggi commented on app/controllers/reservations_controller.rb:
```diff
@@ -295,10 +304,9 @@ def ability_resource
-    @reservation.actual_end_at ||
-    !(@reservation.can_edit? || @reservation.can_customer_edit?)
+    different_id = params[:id].to_i != @reservation.id
+    cannot_edit = current_user.administrator? ? !@reservation.can_edit? : !@reservation.can_customer_edit?
```
I'm not sure this check needs to change. Also, we want to be careful with how we talk about "administrators". `User#administer?` is asking if the user is a global admin. Usually, we would want to use some other check like `manager_of` or something in the cancan abilities.

jdzak replied:
I removed  the check for `@reservation.actual_start_at` because if a reservation has been started, then it will have `@reservation.actual_start_at` which would make `invalid_for_update?` return true.

I can probably re-add the check for `@reservation.actual_end_at` because we want to keep that.

For the abilities check, maybe we can work through this together because I want to make sure I do it right.

jhanggi commented on app/models/reservation.rb:
```diff
@@ -189,7 +189,22 @@ def can_cancel?
+  end
+
+  def reserve_start_at_editable?
+    Time.zone.now < reserve_start_at && before_lock_window?
```
This feels like it might be redundant. Wouldn't `before_lock_window?` always be true when `Time.zone.now < reserve_start_at`?

jdzak replied:
I don't think so because if the lock window is 15 minutes before the reservation, then this condition would be false for the 15 minute block before the reservation starts. Right?

jhanggi commented on app/support/reservations/duration_change_validations.rb:
```diff
@@ -0,0 +1,32 @@
+    @reservation = reservation
+  end
+
+  def reservation
```
Not a big deal, but I'm partial to `attr_reader` for this.

jdzak replied:
Me too, I'll change it.

jhanggi commented on app/support/reservations/duration_change_validations.rb:
```diff
@@ -0,0 +1,32 @@
+    @reservation
+  end
+
+  def copy_errors!
```
Do we have access to `after_validation` that we could hook this method into? I'd like to see this happen automatically after running the validations. Maybe overriding `valid?` with a call to `super` in there.

jdzak replied:
Yep, i'll change it.

jhanggi commented on app/support/reservations/duration_change_validations.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def start_time_not_changed
+    if reservation.reserve_start_at_changed?
+      errors.add(:reserve_start_at, "cannot change once the reservation has started")
```
Can we put the text into I18n?

jdzak replied:
Sure.

jhanggi commented on app/support/reservations/duration_change_validations.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  def start_time_not_changed
+    if reservation.reserve_start_at_changed?
```
Would it make sense to move the check we're doing in the Proc into here? Would that be easier to read?

jdzak replied:
Yea, I was on the fence about this. Having the proc with the if/unless feels nice because we can scroll through the list of validations and see the preconditions for when each validation is fired.

IDK, I don't feel strongly, so if you prefer the other way I can change it.

jhanggi commented on app/support/reservations/duration_change_validations.rb:
```diff
@@ -0,0 +1,32 @@
+    end
+  end
+
+  def duration_not_shortened
```
Same notes as the other method.

jhanggi commented on app/views/reservations/_reservation_fields.html.haml:
```diff
@@ -1,8 +1,20 @@
+      - if start_editable
+        = time_select f, :reserve_start, minute_step: @instrument.reserve_interval
+      - else
+        .controls
```
Can we have `.controls` inside of `.controls`?

jdzak replied:
I'm not sure, I've been experimenting with the styles because I can't get it to vertically align how i want. I may need some help with this too.

jhanggi commented on spec/app_support/reservations/duration_change_validations_spec.rb:
```diff
@@ -0,0 +1,91 @@
+describe Reservations::DurationChangeValidations do
+
+  subject(:validator) { described_class.new(reservation) }
+  let(:reservation) { create :setup_reservation }
```
Can these tests be run with `build_stubbed`?

jdzak replied:
I tried but I got the error:
```
     Failure/Error: let(:reservation) { build_stubbed :setup_reservation }
     ActiveRecord::InvalidForeignKey:
       Mysql2::Error: Cannot add or update a child row: a foreign key constraint fails (`nucore-open_test`.`schedules`, CONSTRAINT `fk_schedules_facility` FOREIGN KEY (`facility_id`) REFERENCES `facilities` (`id`)): INSERT INTO `schedules` (`created_at`, `facility_id`, `name`, `updated_at`) VALUES ('2015-09-11 14:30:00', 1001, 'Schedule 1', '2015-09-11 14:30:00')
     # ./spec/factories/products.rb:81:in `block (3 levels) in <top (required)>'
     # ./spec/app_support/reservations/duration_change_validations_spec.rb:6:in `block (2 levels) in <top (required)>'
     # ./spec/app_support/reservations/duration_change_validations_spec.rb:5:in `block (2 levels) in <top (required)>'
     # ./spec/app_support/reservations/duration_change_validations_spec.rb:42:in `block (3 levels) in <top (required)>'
```

jhanggi commented on spec/app_support/reservations/duration_change_validations_spec.rb:
```diff
@@ -0,0 +1,91 @@
+          reserve_end_at: 30.minutes.from_now)
+      end
+
+      it { expect(validator).to be_invalid }
```
It's probably worth checking that the `errors` contains the right thing. That way we're not returning false positives when some other validation is causing it to be invalid.

jdzak replied:
Sure.

jhanggi commented on spec/controllers/reservations_controller_spec.rb:
```diff
@@ -603,7 +603,8 @@
         let(:next_reservation) do
           Reservation.new :reserve_start_at => Time.zone.parse('2013-08-15 12:02'),
-                          :reserve_end_at   => Time.zone.parse('2013-08-15 12:17')
+                          :reserve_end_at   => Time.zone.parse('2013-08-15 12:17'),
```
Can we start switching to 1.9 hash syntax as we're working on new stuff? No need for a full-out campaign, but we should change it where we're touching the code.

jdzak replied:
Sure.

jdzak commented:
Ready for review. 

There's one failing test which I'm still trying to debug, but other than that it's ready to go.

jdzak commented:
Btw the failing test is `rspec spec/controllers/reservations_controller_spec.rb:820` if you were curious in taking a look at it.

jhanggi commented on app/controllers/reservations_controller.rb:
```diff
@@ -295,8 +302,7 @@ def ability_resource
   def invalid_for_update?
-    params[:id].to_i != @reservation.id ||
-    @reservation.actual_start_at ||
+    different_id = params[:id].to_i != @reservation.id
```
Where does `different_id` get used? Should this have a `||` at the end of the line?

jdzak replied:
Whoops, let me fix both of those issues.

jhanggi commented on app/support/reservations/duration_change_validations.rb:
```diff
@@ -0,0 +1,42 @@
+
+  def valid?(context = nil)
+    super(context).tap do
+      copy_errors!
```
Could we use the `after_validation` callback for this?

jdzak replied:
Ya, I'll change it.

jhanggi commented on app/views/reservations/_reservation_fields.html.haml:
```diff
@@ -1,8 +1,15 @@
       .string.optional.control-label &nbsp;
-      = time_select f, :reserve_start, minute_step: @instrument.reserve_interval
+      = time_select f, :reserve_start, { minute_step: @instrument.reserve_interval }, disabled: start_disabled
+      - if start_disabled
```
Is this really necessary? I would expect it to simply not update the start at field if it doesn't get the parameters.

jdzak replied:
Initially I tried to make it work without the hidden fields, but I was running into a issues because `Reservation#assign_times_from_params` will [set Reservation#reserve_start_at and Reservation#reserve_end_at to nil so the dates are set correctly](https://github.com/tablexi/nucore-open/blob/master/app/support/reservations/date_support.rb#L177-L186). Alternatively, I could probably fill in reserve_start_at and reserve_end_at in the controller if they don't exist in the params so we can remove the hidden fields from the page.

jhanggi replied:
Does this relate at all to the `reservation_params.tap` up in the controller? With that, could we get rid of this?

jdzak replied:
Yep.

jdzak commented on spec/controllers/reservations_controller_spec.rb:
```diff
@@ -618,7 +619,8 @@
         let(:next_reservation) do
           Reservation.new :reserve_start_at => Time.zone.parse('2013-08-15 12:05:30'),
-                          :reserve_end_at   => Time.zone.parse('2013-08-15 12:20:30')
+                          :reserve_end_at   => Time.zone.parse('2013-08-15 12:20:30'),
```
Let me convert these to the 1.9 hash syntax also.

jdzak commented:
Ready for re-review.

jhanggi commented on app/controllers/reservations_controller.rb:
```diff
@@ -295,10 +302,12 @@ def ability_resource
-    @reservation.actual_start_at ||
     @reservation.actual_end_at ||
-    !(@reservation.can_edit? || @reservation.can_customer_edit?)
+    @reservation.actual_start_at ||
```
I believe this is now preventing me from editing a started reservation.

jhanggi commented:
I removed that one line to try to do some further testing.

If I try to shorten the time, I get the "Reserve end at cannot be shortened once the reservation has started" message, but now the duration is set to the shortened time and is disabled. I also see a similar thing when my lengthened reservation conflicts with another reservation.

jdzak commented:
Weird. It wasn't doing that before. I was tinkering with
`invalid_for_update?` recently, I wonder if I screwed something up with
that. I'll take a look at it tonight.

On Wed, Apr 22, 2015 at 3:42 PM, Jason Hanggi <notifications@github.com>
wrote:

> I removed that one line to try to do some further testing.
>
> If I try to shorten the time, I get the "Reserve end at cannot be
> shortened once the reservation has started" message, but now the duration
> is set to the shortened time and is disabled. I also see a similar thing
> when my lengthened reservation conflicts with another reservation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tablexi/nucore-open/pull/271#issuecomment-95328709>.
>

jdzak commented:
Alright, fixed the issue. I reverted some of the `ReservationController#invalid_for_update?` code and forgot to re-test with relay reservations. Also, I'm not enabling/disabling the start/end time fields based on the original `Reservation` record and not the un-persisted user entered values.

jdzak commented:
Also, Todd said we should enforce the lock window before a reservation has started, so I added code for that.

jhanggi commented:
It's 1:08pm. I "Create & Start" a reservation beginning at 1:10pm.

Current behavior:  For the next two minutes, I have the ability to change the reservation start time. Once 1:10 hits, it locks the start_at.

Expected behavior: Once the reservation has actually started, I should not be able to edit the start time.

jhanggi commented:
If the start time is locked, I can edit the hidden fields (not a problem). When I submit, it is changing the reserve_start_at of the reservation. If the start at is locked, we should throw away any of the start at fields in the controller.

jdzak commented:
Alright, fixed those two issues. Ready for re-review.

jhanggi commented on app/controllers/reservations_controller.rb:
```diff
@@ -183,7 +183,24 @@ def update
     end
 
-    @reservation.assign_times_from_params(params[:reservation])
+    reservation_params = params[:reservation]
```
What if we took this entire chunk and extracted it into a `reservation_params` method?

jdzak replied:
I like, that'll look a lot cleaner.

jhanggi commented on app/controllers/reservations_controller.rb:
```diff
@@ -183,7 +183,24 @@ def update
+    @reservation.assign_times_from_params(reservation_params)
+
+    validator = Reservations::DurationChangeValidations.new(@reservation)
+    if validator.invalid?
```
This might also be a candidate for extracting the method.

jhanggi replied:
On thing that might be nice is to make a `render_edit` method that does `set_window` and `render :edit`. Then you could use one method here and at the bottom of `update`.

So this would become something like
```ruby
render_update and return unless duration_change_valid?
```
then
```ruby
def duration_change_valid?
   Reservations::DurationChangeValidations.new(@reservation).valid?
end
```

It might be worth reworking `invalid_for_update?` to be a positive. Not a big deal, I generally prefer positive method names than negative. Unless that leads you to try something like `unless/else`. That way lies the path to madness.

jdzak replied:
LOL. Makes sense, I'll change this.

jhanggi commented on app/helpers/reservations_helper.rb:
```diff
@@ -27,4 +27,18 @@ def reservation_actions(reservation)
     ReservationUserActionPresenter.new(self, reservation).view_edit_link
   end
+
+  def start_time_disabled(reservation)
```
Since these methods return booleans, perhaps they should end with `?`.

jdzak replied:
Sure.

jhanggi commented on app/models/reservation.rb:
```diff
@@ -189,13 +189,32 @@ def can_cancel?
+    outside_lock_window? && Time.zone.now <= reserve_end_at && next_duration_available? && actual_end_at.blank?
+  end
+
+  def next_duration_available?
```
Is this at all related to `ReservationFinder#next_reservation` in Product::SchedulingSupport? It's a really confusing method, but it might do something similar to what you're looking for.

We should make sure that you also can't extend your reservation outside of the schedule rules. E.g. There's a schedule rule for 9-5, and you have a reservation from 4-5. You shouldn't be able to extend beyond that.

jdzak replied:
Yea, I'll switch to use that instead and add a spec for the end of day situation.

jdzak commented:
Ready for re-review.

jhanggi commented:
:+1: It might be interesting to extract some of the `can_customer_edit?` and `reserve_start_at_editable?` type methods into a policy/permissions class, but I'm not sure if it's worth the effort at this point, but if we end up having to play more with this feature, it might be worth considering.

Also, we'll wait until we get the relay features released to prod before we merge this down.

jhanggi commented:
I ran this on CI, and it has some failures.

jdzak commented:
Fixed the spec failures. It was failing because I added some code to check if `actual_start_at` exists on the reservation, but accidentally removed the check to see if `reserve_start_at check` exists.
